### ISSUE TITLE
[MNG-5561] Plugin relocation loses configuration

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6777]
**[Julien CARSIQUE](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jcarsique)** opened **[MNG-5561](https://issues.apache.org/jira/browse/MNG-5561?redirect=false)** and commented

Hi,

Applying a relocation on a plugin seems to lose its configuration.
In the following sample, I relocated "nuxeo-distribution-tools" to "ant-assembly-maven-plugin" and the missing configured parameter is "buildFile". The relocation itself went well but the execution fails because of a missing parameter.

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>org.nuxeo.build</groupId>
  <artifactId>nuxeo-distribution-tools</artifactId>
  <version>2.0-SNAPSHOT</version>
  <packaging>maven-plugin</packaging>
  <name>Nuxeo Distribution Build</name>
  <description>Assemble Nuxeo distributions</description>
  <distributionManagement>
    <relocation>
      <artifactId>ant-assembly-maven-plugin</artifactId>
    </relocation>
  </distributionManagement>
</project>
```

```
<build>
  <plugins>
    <plugin>
      <groupId>org.nuxeo.build</groupId>
      <!-- With relocation -->
      <artifactId>nuxeo-distribution-tools</artifactId>
      <!-- Without relocation -->
      <!-- <artifactId>ant-assembly-maven-plugin</artifactId> -->
      <executions>
        <execution>
          <id>assemble</id>
          <phase>package</phase>
          <configuration>
            <!-- The missing parameter in case of relocation -->
            <buildFile>src/main/assemble/assembly.xml</buildFile>
          </configuration>
          <goals>
            <goal>build</goal>
          </goals>
        </execution>
      </executions>
    </plugin>
  </plugins>
</build>
```

Here is the diff between an "mvn package -X" with nuxeo-distribution-tools (ie: with relocation; line starting with "<") and another one with ant-assembly-maven-plugin (ie: without relocation; line starting with ">"):

```
< [DEBUG] Dependency collection stats: {ConflictMarker.analyzeTime=1, ConflictMarker.markTime=0, ConflictMarker.nodeCount=72, ConflictIdSorter.graphTime=0, ConflictIdSorter.topsortTime=0, ConflictIdSorter.conflictIdCount=34, ConflictIdSorter.conflictIdCycleCount=0, ConflictResolver.totalTime=1, ConflictResolver.conflictItemCount=71, DefaultDependencyCollector.collectTime=10, DefaultDependencyCollector.transformTime=2}
> [DEBUG] Dependency collection stats: {ConflictMarker.analyzeTime=1, ConflictMarker.markTime=0, ConflictMarker.nodeCount=72, ConflictIdSorter.graphTime=0, ConflictIdSorter.topsortTime=0, ConflictIdSorter.conflictIdCount=34, ConflictIdSorter.conflictIdCycleCount=0, ConflictResolver.totalTime=2, ConflictResolver.conflictItemCount=71, DefaultDependencyCollector.collectTime=65, DefaultDependencyCollector.transformTime=3}
&mdash;
< [DEBUG] Created new class realm plugin>org.nuxeo.build:nuxeo-distribution-tools:2.0-SNAPSHOT
< [DEBUG] Importing foreign packages into class realm plugin>org.nuxeo.build:nuxeo-distribution-tools:2.0-SNAPSHOT
> [DEBUG] Created new class realm plugin>org.nuxeo.build:ant-assembly-maven-plugin:2.0-SNAPSHOT
> [DEBUG] Importing foreign packages into class realm plugin>org.nuxeo.build:ant-assembly-maven-plugin:2.0-SNAPSHOT
&mdash;
< [DEBUG] Populating class realm plugin>org.nuxeo.build:nuxeo-distribution-tools:2.0-SNAPSHOT
> [DEBUG] Populating class realm plugin>org.nuxeo.build:ant-assembly-maven-plugin:2.0-SNAPSHOT
&mdash;
< [DEBUG] Configuring mojo org.nuxeo.build:ant-assembly-maven-plugin:2.0-SNAPSHOT:build from plugin realm ClassRealm[plugin>org.nuxeo.build:nuxeo-distribution-tools:2.0-SNAPSHOT, parent: sun.misc.Launcher$AppClassLoader@591ce4fe]
> [DEBUG] Configuring mojo org.nuxeo.build:ant-assembly-maven-plugin:2.0-SNAPSHOT:build from plugin realm ClassRealm[plugin>org.nuxeo.build:ant-assembly-maven-plugin:2.0-SNAPSHOT, parent: sun.misc.Launcher$AppClassLoader@77fe0d66]
&mdash;
> [DEBUG]   (f) buildFile = /home/julien/ws-ssd/nuxeo/nuxeo-distribution/nuxeo-distribution-resources/src/main/assemble/assembly.xml
```

The last line highlights the missing parameter.

Indeed, it seems that in case of a relocation, all the "Parameter" Mojo fields take their default value (or none) whereas the "Component" fields are still properly injected.

---

**Affects:** 3.1.1

**Attachments:**
- [screenshot-1.png](https://issues.apache.org/jira/secure/attachment/13037938/screenshot-1.png) (_48.17 kB_)

**Issue Links:**
- [MNG-7204](https://issues.apache.org/jira/browse/MNG-7204) Maven might pick up the wrong configuration when the plugin happens to include relocation information
 (_**"is duplicated by"**_)
- [MNG-7374](https://issues.apache.org/jira/browse/MNG-7374) Mutating RelocatedArtifact does not retain type

**Remote Links:**
- [GitHub Pull Request #132
](https://github.com/apache/maven-integration-testing/pull/132)
- [GitHub Pull Request #642
](https://github.com/apache/maven/pull/642)
- [GitHub Pull Request #643
](https://github.com/apache/maven/pull/643)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.5](https://github.com/apache/maven/milestone/106?closed=1)

3 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Julien CARSIQUE](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jcarsique)** commented

Reproduction test case:

```
git clone http://github.com/nuxeo/ant-assembly-maven-plugin
cd ant-assembly-maven-plugin
git checkout release-2.0
mvn clean integration-test -DskipTests=true -DdebugITs=true -Dinvoker.test=test1,test-MNG-5561
```

Both tests are identical except that test-MNG-5561 uses nuxeo-distribution-tools relocated to ant-assembly-maven-plugin.
test-MNG-5561 fails with a `FileNotFoundException: .../target/it/test-MNG-5561/build.xml (Aucun fichier ou dossier de ce type)` since the `build.xml` file is expected to be at `.../target/it/test-MNG-5561/src/build.xml` as stated by the `buildFile` configuration parameter.
We can see that with the relocation, the default value is used for the parameter instead of the one configured in the POM.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

Please read: https://cwiki.apache.org/confluence/display/MAVEN/Submitting+Issue+for+Maven's+Core

**Comment by jira-importer** [comment]
**[Julien CARSIQUE](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jcarsique)** commented

Done. I've updated the above comment since I simplified the reproduction case.

I'm not sure to understand what you want.
The reproduction case is the integration test "test-MNG-5561" included in http://github.com/nuxeo/ant-assembly-maven-plugin. Did you encounter any issue while running it?
You wrote that << The preferred way of providing an example project is Github repository >>. Do you want me to create an almost empty project raising the same issue?

**Comment by jira-importer** [comment]
**[Julien CARSIQUE](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jcarsique)** commented

The issue is solved when setting the target parameter as a "User property"

```
-    @Parameter
+    @Parameter(property = "buildFile")
     protected File buildFile;
```

It would mean that a plugin parameter set on the relocated plugin is transmitted to the target plugin as a "User property" instead of a plugin parameter...
I was about to close the issue as "Not a bug" but finally, it still sounds like a bug. What do you think?

EDIT: updated the above reproduction instructions to checkout "release-2.0" tag since the issue is not really solved but hidden by the user property workaround.

**Comment by jira-importer** [comment]
**[Peter Van den Bosch](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peter.vandenbosch@ksz-bcss.fgov.be)** commented

Hi,

I'm also experiencing this issue for a relocated maven plugin. The workaround of assigning a property name to the parameter doesn't seem to work for complex or multi-valued properties. Is there another way to make it work?

```
[ERROR] Failed to execute goal be.fgov.kszbcss.plugins:websphere-maven-plugin:2.8.14-SNAPSHOT:enable-endpoints (default) on project BpEar: No children found in ejbModules tag -> [Help 1]
```

```
@Parameter (property="enable-endpoints.ejbmodules", required=true)
private EJBModule[] ejbModules;
```

```
<plugin>
	<groupId>be.fgov.kszbcss.plugins</groupId>
	<artifactId>maven-websphere-plugin</artifactId>
	<executions>
		<execution>
			<goals>
				<goal>enable-endpoints</goal>
			</goals>
		</execution>
	</executions>
	<configuration>
		<ejbModules>
			<ejbModule>
				<artifactId>BpRequestHandlerEjb</artifactId>
				<contextRoot>/Bp</contextRoot>
				<localNames>
					<PullRequestHandlerHttpPort>/PullRequestHandlerHttp</PullRequestHandlerHttpPort>
				</localNames>
			</ejbModule>
		</ejbModules>
	</configuration>
</plugin>
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Analyzed,

Now, as far as I can see the bug is here:
!screenshot-1.png!

`mojoExecution` contains the **new** GA, but the project contains the plugin under the **old** GA, same for plugin management. Therefore, the subsequent block is never executed and you run a virgin/vanilla plugin.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Digging a bit deeper: The code as such existed eversince, more than 12 years now. I do not really understand why it is necessary to find the plugin in the current project when it is available from `mojoExecution`, but on needs a two stage approach: the one done above and one from `MojoExectuion#getPlugin()` which is the original element from the POM and not the relocated one.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I think I have a solution for the problem which uncovers a dynamically replaced vs statically configured issue. Will share soon.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I have now created now a  PR for which fixes the issue for me. What I don't know for sure whether we can completely rely on `Plugin` or need to fiddle with `MojoExecution` first and then try `Plugin`.
`MojoExecution` represents full dynamic resolution **after** relocation, so it already represents the new plugin without any connection to the old one while `Plugin` contains the old plugin referenced in the POM. Obviously, the new plugin is **not** configured in the POM, but only the old one. I think that we need to read out the static one from POM to find config. This won't matter for plugins which haven't been relocated, but makes complete difference for relocated ones.

[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas), [Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl), [Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte), opinions? I will try to work on an IT as well.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [2670c007517d8d72c182df6e589fd9dc43692e26](https://gitbox.apache.org/repos/asf?p=maven.git&a=commit&h=2670c007517d8d72c182df6e589fd9dc43692e26) in master, with [05b748ff6aa15aa63a9c9d5f9f5679f47bf9e83d](https://gitbox.apache.org/repos/asf?p=maven.git&a=commit&h=05b748ff6aa15aa63a9c9d5f9f5679f47bf9e83d) in 3.8.x and ITs with [e06a28840f202d244188177cd7e24fa41d29061f](https://gitbox.apache.org/repos/asf?p=maven-integration-testing.git&a=commit&h=e06a28840f202d244188177cd7e24fa41d29061f).
